### ISSUE TITLE
Using ctx.SetState when Entity State doesn't exist causes that operation to skip

### ISSUE THREAD

**Issue by raggupta-ms** [Azure/azure-functions-durable-extension#1266]
### Description
This is my code for Entity Function

```
[FunctionName(nameof(Circuit))]
        public static Task Run(
            [EntityTrigger] IDurableEntityContext ctx)
        {
            if (!ctx.HasState)
            {
                // Entity state doesn't exist. This is the first call ever, set startup state.
                ctx.SetState(CircuitEntity.GetFirstStartupState());
            }

            return ctx.DispatchAsync<Circuit>();
        }
```

The first signal call to Entity Function, when the entity doesn't exist (i.e. ctx.HasState = false), sets the state using ctx.SetState. But as a result, the Signal itself is skipped. ctx.Dispatch doesn't trigger my code for that operation.
If I don't set default state for entity, it works fine as expected.

### Expected behavior
Regardless of setting first state of entity, the signal should invoke the entity oprtation.

### Actual behavior
The operation that sets the default state doesn't trigger the operation on entity.

### Relevant source code snippets
```
See above.
```

### Known workarounds
Not that I know of. I need to set default entity state without skipping any operation signal.

### App Details

**Durable Functions extension version (e.g. v1.8.3):**
Microsoft.Azure.Functions.Extensions = 1.0.0
Microsoft.Azure.WebJobs.Extensions.DurableTask = 2.1.1
Microsoft.Azure.WebJobs.Extensions.EventHubs = 4.1.1
Microsoft.NET.Sdk.Functions = 3.0.3

**Azure Functions runtime version (1.0 or 2.0):**
v3

**Programming language used:**
C# (.Net Core 3.1)

### Screenshots
> If applicable, add screenshots to help explain your problem.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**:
- **Function App name**:
- **Function name(s)**:
- **Azure region**:
- **Orchestration instance ID(s)**:
- **Azure storage account name**:

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by sebastianburckhardt** [comment]
I have tested this in my environment and it works as expected. 

What I suspect is happening in your situation is that there is an exception thrown somewhere.
For example, it could be a type error. If you call `ctx.SetState(x)` for some x that is not of type `CircuitEntity`, then `DispatchAsync` fails with an `InvalidCastException`.

Did you check the logs to see if there were any exceptions reported?

**Comment by raggupta-ms** [comment]
Yes you are right. I was trying to set the base class object (Circuit inherits CircuitEntity) as Entity state and that is throwing InvalidCastException. Obviously its a code bug but following is the reason why I never thought it could be a failure and looked into logs -

Before calling ctx.SetState(x), I read the Entity state and it returns 404 (expected). After ctx.SetState(x) (with InvalidCastException), it returns the Entity state successfully. And the values of state didn't raise any suspicion to me because in my explicit default state set call, I had the exact same values :). So it appeared a successful call and I never looked into the logs. 

Setting up Entity state, the very first time, even on a failure leads to confusion. Though it could be the expected behavior as per design but I would have liked it to still return 404 after InvalidCastException.

@sebastianburckhardt Thanks for looking into this.

**Comment by sebastianburckhardt** [comment]
> Setting up Entity state, the very first time, even on a failure leads to confusion. Though it could be the expected behavior as per design but I would have liked it to still return 404 after InvalidCastException.

Yes, I agree. We have recently changed this behavior (see #1231, shipped in v2.2.0): when there is an exception we roll back all changes to the entity, so in this example the entity would not be created, and you would get a 404 afterwards .
